assignment/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java

```diff
@@ -16757,4 +16773,12 @@ private String getRateSubmissionTimeSpent(AssignmentSubmission submission) {
         }
         return assignmentService.intToTime(assigmentRateSpent);
     }
+    
+    private Optional<User> getUser(String userId) {
```

Comment:
You could move this into the UserDirectoryService. Maybe use the method "getOptionalUser"? We could perhaps use that naming pattern as a consistent way of overloading existing methods like getUser.

Reply:
Yeah, I think it makes sense, but I didn't want to go so far. I'll do it.

Reply:
Maybe we could keep it inline in the stream, it's a tad uglier but doesn't need kernel updates

Reply:
I have a special "Master adjustments" commit on this branch, I would be fine with adding it there as well. Because I think this pattern makes sense and will be introduced at some point anyway. If we anticipate it now, we won't have duplicated code later. For 22 based branched `this::getOptionalUser` and master `userDirectoyService::getOptionalUser.` What do you think?

Reply:
It's fine. I would rephrase it as "Master master adjustments" though

Reply:
Agree with everyone here that adding a new method UserDirectoyService::getOptionalUser to return an Optional aka
`public Optional<User> getOptionalUser(String userId)`

I approve of the method name here, and the original method should be annotated as deprecated. (it's a shame that the return type is not part of a method signature)

this also has a positive side effect that the Exception UserNotDefinedException does not need to be thrown. TBH this was what returning NULL or Optional<> is for. An Exception is for communicating to the callee of the method with a return value other than Objects presence or absence is not to be trusted.

Reply:
I did add the new method to UserDirectoryService, but I did not deprecate the original one yet